File: frontend/src/lib/api/extcommunity-list.ts
import { apiClient } from "./client";

// ============================================================================
// TypeScript Interfaces
// ============================================================================

export interface ExtCommunityListRule {
  rule_number: number;
  description?: string | null;
  action: string;  // permit|deny
  regex?: string | null;
}

export interface ExtCommunityList {
  name: string;
  description?: string | null;
  rules: ExtCommunityListRule[];
}

export interface ExtCommunityListConfig {
  extcommunity_lists: ExtCommunityList[];
  total: number;
}

export interface ExtCommunityListCapabilities {
  version: string;
  features: {
    basic: { supported: boolean; description: string };
    rules: { supported: boolean; description: string };
    actions: { supported: boolean; description: string };
  };
  version_notes: {
    identical_versions: string;
  };
  device_name?: string;
}

export interface ExtCommunityListBatchOperation {
  op: string;
  value?: string;
}

export interface ExtCommunityListBatchRequest {
  name: string;
  rule_number?: number;
  operations: ExtCommunityListBatchOperation[];
}

// ============================================================================
// API Service
// ============================================================================

class ExtCommunityListService {
  /**
   * Get capabilities based on VyOS version
   */
  async getCapabilities(): Promise<ExtCommunityListCapabilities> {
    return apiClient.get<ExtCommunityListCapabilities>("/vyos/extcommunity-list/capabilities");
  }

  /**
   * Get all extcommunity lists configuration
   */
  async getConfig(refresh: boolean = false): Promise<ExtCommunityListConfig> {
    return apiClient.get<ExtCommunityListConfig>("/vyos/extcommunity-list/config", {
      refresh: refresh.toString(),
    });
  }

  /**
   * Refresh the cached configuration
   */
  async refreshConfig(): Promise<unknown> {
    return apiClient.post("/vyos/config/refresh");
  }

  /**
   * Execute batch operations
   */
  async batchConfigure(request: ExtCommunityListBatchRequest): Promise<unknown> {
    const result = await apiClient.post("/vyos/extcommunity-list/batch", request);
    await this.refreshConfig();
    return result;
  }

  /**
   * Delete an entire extcommunity list
   */
  async deleteExtCommunityList(name: string): Promise<unknown> {
    const operations: ExtCommunityListBatchOperation[] = [];
    operations.push({ op: "delete_extcommunity_list" });

    return this.batchConfigure({
      name,
      rule_number: 0, // Not used for delete_extcommunity_list
      operations,
    });
  }

  /**
   * Delete a specific rule from a extcommunity list and renumber remaining rules to close gaps
   */
  async deleteRule(name: string, ruleNumber: number): Promise<unknown> {
    // Get current configuration
    const config = await this.getConfig(true);
    const communityList = config.extcommunity_lists.find(cl => cl.name === name);

    if (!communityList) {
      throw new Error(`ExtCommunity list ${name} not found`);
    }

    // Get remaining rules (excluding the one being deleted)
    const remainingRules = communityList.rules.filter(r => r.rule_number !== ruleNumber);

    if (remainingRules.length === 0) {
      // If no rules left, just delete the rule
      const operations: ExtCommunityListBatchOperation[] = [];
      operations.push({ op: "delete_rule" });

      return this.batchConfigure({
        name,
        rule_number: ruleNumber,
        operations,
      });
    }

    // Sort remaining rules by their current number
    const sortedRules = remainingRules.sort((a, b) => a.rule_number - b.rule_number);

    // Renumber sequentially starting from the lowest existing number
    // This closes gaps: if you have 105, 106, 107, 108 and delete 106,
    // result will be 105, 106 (was 107), 107 (was 108)
    const startingNumber = sortedRules[0].rule_number;
    const reorderRules = sortedRules.map((rule, index) => ({
      old_number: rule.rule_number,
      new_number: startingNumber + index,
      rule_data: rule,
    }));

    // Check if any rule actually needs renumbering
    const needsReorder = reorderRules.some(r => r.old_number !== r.new_number);

    if (!needsReorder) {
      // No gaps to close, just delete the rule directly
      const operations: ExtCommunityListBatchOperation[] = [];
      operations.push({ op: "delete_rule" });

      return this.batchConfigure({
        name,
        rule_number: ruleNumber,
        operations,
      });
    }

    // Use reorder endpoint which will delete all rules (including the one we want to delete)
    // and recreate them with new sequential numbers
    return this.reorderRules(name, reorderRules);
  }

  /**
   * Helper: Create a new extcommunity list with a rule
   */
  async createExtCommunityList(name: string, description: string | null, rule: Partial<ExtCommunityListRule>): Promise<unknown> {
    const operations: ExtCommunityListBatchOperation[] = [];

    // Create extcommunity list
    operations.push({ op: "set_extcommunity_list" });

    // Add description
    if (description) {
      operations.push({ op: "set_extcommunity_list_description", value: description });
    }

    // Create rule
    operations.push({ op: "set_rule" });

    // Rule description
    if (rule.description) {
      operations.push({ op: "set_rule_description", value: rule.description });
    }

    // Rule action
    if (rule.action) {
      operations.push({ op: "set_rule_action", value: rule.action });
    }

    // Rule regex
    if (rule.regex) {
      operations.push({ op: "set_rule_regex", value: rule.regex });
    }

    return this.batchConfigure({
      name,
      rule_number: rule.rule_number,
      operations,
    });
  }

  /**
   * Helper: Update an existing extcommunity list or rule
   */
  async updateExtCommunityList(
    name: string,
    originalExtCommunityList: ExtCommunityList,
    description: string | null,
    rule?: Partial<ExtCommunityListRule>,
    ruleNumber?: number
  ): Promise<unknown> {
    const operations: ExtCommunityListBatchOperation[] = [];

    // Update description
    if (description !== originalExtCommunityList.description) {
      if (description) {
        operations.push({ op: "set_extcommunity_list_description", value: description });
      } else if (originalExtCommunityList.description) {
        operations.push({ op: "delete_extcommunity_list_description" });
      }
    }

    // If updating a rule
    if (rule && ruleNumber !== undefined) {
      const originalRule = originalExtCommunityList.rules.find(r => r.rule_number === ruleNumber);

      // Rule description
      if (rule.description !== originalRule?.description) {
        if (rule.description) {
          operations.push({ op: "set_rule_description", value: rule.description });
        } else if (originalRule?.description) {
          operations.push({ op: "delete_rule_description" });
        }
      }

      // Rule action
      if (rule.action && rule.action !== originalRule?.action) {
        operations.push({ op: "set_rule_action", value: rule.action });
      }

      // Rule regex
      if (rule.regex !== originalRule?.regex) {
        if (rule.regex) {
          operations.push({ op: "set_rule_regex", value: rule.regex });
        } else if (originalRule?.regex) {
          operations.push({ op: "delete_rule_regex" });
        }
      }
    }

    return this.batchConfigure({
      name,
      rule_number: ruleNumber,
      operations,
    });
  }

  /**
   * Helper: Add a new rule to existing extcommunity list
   */
  async addRule(name: string, rule: Partial<ExtCommunityListRule>): Promise<unknown> {
    const operations: ExtCommunityListBatchOperation[] = [];

    // Create rule
    operations.push({ op: "set_rule" });

    // Rule description
    if (rule.description) {
      operations.push({ op: "set_rule_description", value: rule.description });
    }

    // Rule action
    if (rule.action) {
      operations.push({ op: "set_rule_action", value: rule.action });
    }

    // Rule regex
    if (rule.regex) {
      operations.push({ op: "set_rule_regex", value: rule.regex });
    }

    return this.batchConfigure({
      name,
      rule_number: rule.rule_number,
      operations,
    });
  }

  /**
   * Update an existing rule
   */
  async updateRule(name: string, ruleNumber: number, rule: Partial<ExtCommunityListRule>): Promise<unknown> {
    const operations: ExtCommunityListBatchOperation[] = [];

    // Rule description
    if (rule.description !== undefined) {
      if (rule.description) {
        operations.push({ op: "set_rule_description", value: rule.description });
      } else {
        operations.push({ op: "delete_rule_description" });
      }
    }

    // Rule action
    if (rule.action) {
      operations.push({ op: "set_rule_action", value: rule.action });
    }

    // Rule regex
    if (rule.regex !== undefined) {
      if (rule.regex) {
        operations.push({ op: "set_rule_regex", value: rule.regex });
      } else {
        operations.push({ op: "delete_rule_regex" });
      }
    }

    return this.batchConfigure({
      name,
      rule_number: ruleNumber,
      operations,
    });
  }

  /**
   * Reorder extcommunity list rules
   */
  async reorderRules(extcommunityListName: string, rules: Array<{ old_number: number; new_number: number; rule_data: ExtCommunityListRule }>): Promise<unknown> {
    const result = await apiClient.post("/vyos/extcommunity-list/reorder", {
      extcommunity_list_name: extcommunityListName,
      rules: rules,
    });
    await this.refreshConfig();
    return result;
  }
}

export const extcommunityListService = new ExtCommunityListService();

File: backend/routers/extcommunity_list/extcommunity_list.py
"""
ExtExtCommunity List Router

API endpoints for managing VyOS extcommunity-list configuration.
Supports version-aware configuration for VyOS 1.4 and 1.5 (identical feature sets).
"""

from fastapi import APIRouter, HTTPException, Request
from starlette.concurrency import run_in_threadpool
from pydantic import BaseModel, Field
from typing import List, Dict, Optional, Any
from session_vyos_service import get_session_vyos_service
from vyos_builders import ExtCommunityListBatchBuilder
from fastapi_permissions import require_read_permission, require_write_permission
from rbac_permissions import FeatureGroup
from utils.router_helpers import load_vyos_capabilities
import inspect

router = APIRouter(prefix="/vyos/extcommunity-list", tags=["extcommunity-list"])

# Stub functions for backwards compatibility with app.py
def set_device_registry(registry):
    """Legacy function - no longer used."""
    pass


def set_configured_device_name(name):
    """Legacy function - no longer used."""
    pass


# ============================================================================
# Pydantic Models
# ============================================================================


class ExtCommunityListRule(BaseModel):
    """ExtExtCommunity list rule"""
    rule_number: int
    description: Optional[str] = None
    action: str = "permit"  # permit|deny
    regex: Optional[str] = None


class ExtCommunityList(BaseModel):
    """Complete extcommunity list configuration"""
    name: str
    description: Optional[str] = None
    rules: List[ExtCommunityListRule] = []


class ExtCommunityListConfig(BaseModel):
    """Response containing all extcommunity lists"""
    extcommunity_lists: List[ExtCommunityList] = []
    total: int = 0


class ExtCommunityListBatchOperation(BaseModel):
    """Single operation in a batch request"""
    op: str = Field(..., description="Operation name")
    value: Optional[str] = Field(None, description="Operation value")


class ExtCommunityListBatchRequest(BaseModel):
    """Model for batch configuration"""
    name: str = Field(..., description="ExtExtCommunity list name")
    rule_number: Optional[int] = Field(None, description="Rule number (optional)")
    operations: List[ExtCommunityListBatchOperation]


class ReorderRuleItem(BaseModel):
    """Single rule in a reorder request"""
    old_number: int = Field(..., description="Original rule number")
    new_number: int = Field(..., description="New rule number after reorder")
    rule_data: ExtCommunityListRule = Field(..., description="Complete rule configuration")


class ReorderExtCommunityListRequest(BaseModel):
    """Model for reordering extcommunity list rules"""
    extcommunity_list_name: str = Field(..., description="ExtExtCommunity list name")
    rules: List[ReorderRuleItem] = Field(..., description="List of rules with new order")


class VyOSResponse(BaseModel):
    """Standard response from VyOS operations"""
    success: bool
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


# ============================================================================
# Endpoint 1: Capabilities
# ============================================================================


@router.get("/capabilities")
async def get_extcommunity_list_capabilities(request: Request):
    """
    Get feature capabilities based on device VyOS version.

    Returns feature flags indicating which operations are supported.
    Allows frontends to conditionally enable/disable features.
    """
    return await load_vyos_capabilities(
        request=request,
        feature_group=FeatureGroup.BGP_EXTENDED_COMMUNITY,
        builder_cls=ExtCommunityListBatchBuilder,
    )


# ============================================================================
# Endpoint 2: Config (Generalized Data)
# ============================================================================


@router.get("/config", response_model=ExtCommunityListConfig)
async def get_extcommunity_list_config(http_request: Request, refresh: bool = False):
    """
    Get all extcommunity-list configuration from VyOS in a generalized format.

    Args:
        refresh: If True, force refresh from VyOS. If False, use cache.

    Returns:
        Generalized configuration data optimized for frontend consumption
    """
    try:
        service = get_session_vyos_service(http_request)
        full_config = await run_in_threadpool(service.get_full_config, refresh=refresh)

        # Navigate to policy -> extcommunity-list
        extcommunity_list_config = full_config.get("policy", {}).get("extcommunity-list", {})

        if not extcommunity_list_config:
            return ExtCommunityListConfig(extcommunity_lists=[], total=0)

        # Parse extcommunity lists
        extcommunity_lists = []
        for name, cl_data in extcommunity_list_config.items():
            extcommunity_list = parse_extcommunity_list(name, cl_data)
            extcommunity_lists.append(extcommunity_list)

        return ExtCommunityListConfig(extcommunity_lists=extcommunity_lists, total=len(extcommunity_lists))

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


def parse_extcommunity_list(name: str, cl_data: dict) -> ExtCommunityList:
    """Parse extcommunity list configuration from VyOS format."""
    description = cl_data.get("description")

    rules = []
    rules_raw = cl_data.get("rule", {})
    if rules_raw:
        for rule_num, rule_data in rules_raw.items():
            rule = parse_rule(int(rule_num), rule_data)
            rules.append(rule)

    return ExtCommunityList(
        name=name,
        description=description,
        rules=sorted(rules, key=lambda r: r.rule_number)
    )


def parse_rule(rule_number: int, rule_data: dict) -> ExtCommunityListRule:
    """Parse extcommunity list rule from VyOS format."""
    description = rule_data.get("description")
    action = rule_data.get("action", "permit")
    regex = rule_data.get("regex")

    return ExtCommunityListRule(
        rule_number=rule_number,
        description=description,
        action=action,
        regex=regex
    )


# ============================================================================
# Endpoint 3: Batch Operations
# ============================================================================


@router.post("/batch")
async def extcommunity_list_batch_configure(http_request: Request, body: ExtCommunityListBatchRequest):
    """
    Execute a batch of configuration operations.

    Allows multiple changes in a single VyOS commit for efficiency.
    """
    try:
        service = get_session_vyos_service(http_request)
        version = service.get_version()
        builder = ExtCommunityListBatchBuilder(version=version)

        # Process operations using inspect for dynamic method calls
        for operation in body.operations:
            method = getattr(builder, operation.op)
            sig = inspect.signature(method)
            params = list(sig.parameters.keys())

            # Build arguments dynamically
            args = []

            # Add extcommunity list name
            if "name" in params:
                args.append(body.name)

            # Add rule number if specified and method accepts it
            if body.rule_number and "rule" in params:
                args.append(str(body.rule_number))

            # Add operation value if provided
            if operation.value and len(params) > len(args):
                # Check remaining parameters
                remaining_params = params[len(args):]
                for param in remaining_params:
                    if param != "self":
                        args.append(operation.value)
                        break

            method(*args)

        # Execute batch
        response = service.execute_batch(builder)

        return VyOSResponse(
            success=response.status == 200,
            data={"message": "Configuration updated"},
            error=response.error if response.error else None
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


# ============================================================================
# Reorder Endpoint
# ============================================================================


@router.post("/reorder")
async def reorder_extcommunity_list_rules(http_request: Request, body: ReorderExtCommunityListRequest):
    """
    Reorder extcommunity list rules by deleting and recreating them in a single commit.

    This endpoint efficiently reorders multiple rules by:
    1. Deleting all specified rules in reverse order
    2. Recreating them with new rule numbers
    All operations are executed in a single VyOS commit.

    Args:
        request: Reorder request containing extcommunity list name and list of rules

    Returns:
        VyOSResponse with success/failure information
    """
    try:
        service = get_session_vyos_service(http_request)
        version = service.get_version()
        builder = ExtCommunityListBatchBuilder(version=version)

        # Step 1: Delete all rules in reverse order
        rules_to_delete = sorted([r.old_number for r in body.rules], reverse=True)
        for old_number in rules_to_delete:
            builder.delete_rule(body.extcommunity_list_name, str(old_number))

        # Step 2: Recreate rules with new numbers
        for rule_item in body.rules:
            new_number = rule_item.new_number
            rule_data = rule_item.rule_data

            # Create the rule
            builder.set_rule(body.extcommunity_list_name, str(new_number))

            # Set action
            if rule_data.action:
                builder.set_rule_action(body.extcommunity_list_name, str(new_number), rule_data.action)

            # Set description
            if rule_data.description:
                builder.set_rule_description(body.extcommunity_list_name, str(new_number), rule_data.description)

            # Set regex
            if rule_data.regex:
                builder.set_rule_regex(body.extcommunity_list_name, str(new_number), rule_data.regex)

        # Execute batch
        response = service.execute_batch(builder)

        return VyOSResponse(
            success=response.status == 200,
            data={"message": f"Successfully reordered {len(body.rules)} rules in extcommunity list {body.extcommunity_list_name}"},
            error=response.error if response.error else None
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
